Document self-hosted Sandboxes in additional features guide (#4777)
langchain-ai/docs: src/langsmith/deploy-self-hosted-full-platform.mdx

### Supported platforms

Self-hosted Sandboxes are currently supported on:

- Amazon Elastic Kubernetes Service (EKS)
- Google Kubernetes Engine (GKE)

Azure Kubernetes Service (AKS) is supported by the base LangSmith chart, but self-hosted Sandboxes are not supported on AKS yet.

### Components

Enabling Sandboxes provisions the following resources:

- Sandbox runtime pods that run sandbox workloads on KVM-capable nodes.
- The JuiceFS CSI driver and a JuiceFS-backed volume for sandbox files and snapshots.
- A JuiceFS metadata store backed by Redis and object storage backed by S3 or GCS.
- Optional wildcard ingress for services exposed from inside Sandboxes.

### Prerequisites

<Steps>
  <Step title="Install the base LangSmith platform">
    Install LangSmith on Kubernetes before enabling Sandboxes. See [Self-host LangSmith on Kubernetes](/langsmith/kubernetes).

    Sandboxes run in the same Kubernetes cluster and namespace as the LangSmith release.
  </Step>

  <Step title="Add KVM-capable nodes">
    Your cluster must include dedicated nodes that can run nested workloads with Linux KVM available at `/dev/kvm`.

>     - On AWS, use x86_64 EC2 bare-metal instance types.

Reviewer: AWS also supports nested virtualization now

  enabled: false
```

## Enable Sandboxes

<Info>
Self-hosted Sandboxes require LangSmith Helm chart `0.16.0` or later.
</Info>

Sandboxes are disabled by default. After installation, see [LangSmith Sandboxes](/langsmith/sandboxes) for user workflows in the LangSmith UI and APIs.

### Supported platforms

> Self-hosted Sandboxes are currently supported on:

Reviewer 2: ```suggestion
Self-hosted Sandboxes are supported on:
```

Reviewer 2: Let's remove forward-looking feature language to maintain user expectations and make docs more maintainable

- Amazon Elastic Kubernetes Service (EKS)
- Google Kubernetes Engine (GKE)

> Azure Kubernetes Service (AKS) is supported by the base LangSmith chart, but self-hosted Sandboxes are not supported on AKS yet.

Reviewer 2: ```suggestion
Azure Kubernetes Service (AKS) is supported by the base LangSmith chart, but self-hosted Sandboxes are not supported on AKS.
```

      <Tab title="Using inline values">
        If the Helm chart manages your LangSmith app Secret, set the sandbox secret values directly in your config file. Avoid committing this file to version control.

        ```yaml
        config:
          sandboxes:
            callbackSigningJwk: '<ed25519-private-jwk>'
        ```
      </Tab>
    </Tabs>

    The callback signing value must be an Ed25519 private JWK. Keep it stable across upgrades.
  </Step>

  <Step title="Choose a proxy CA mode">
    The chart supports two proxy CA modes:

    | Mode | Use when |
    | --- | --- |
    | `generatedSecret` | You want Helm to create a self-signed CA Secret. This is the default. |
    | `existingSecret` | You manage the CA Secret outside the LangSmith chart. The Secret can be created manually, by cert-manager, or by another external process. |

    In GitOps workflows that render manifests without live cluster access, prefer `existingSecret`. The `generatedSecret` mode uses Helm's live `lookup` behavior to reuse the generated Secret on upgrades; pure render workflows cannot read the live Secret and may produce new cert material on each render.
  </Step>
</Steps>

### Enable with Helm

> Add the following values to your `langsmith_config.yaml`, along with the sandbox secret values described above. Replace placeholders with your deployment-specific values.

Reviewer 2: ```suggestion
Add the following values to your `langsmith_config.yaml`, along with the sandbox secret values described in the [Prerequisites](#prerequisites-2). Replace placeholders with your deployment-specific values.
```

  enabled: true
  juicefs:
    csi:
      existingSecretName: "juicefs-csi-config"
```

The existing Secret must be in the LangSmith release namespace and contain these keys:

```yaml
stringData:
  name: "sandbox-juicefs"
  metaurl: "redis://redis-host:6379/1"
  storage: "s3"
  bucket: "https://bucket-name.s3.us-west-2.amazonaws.com"
```

Apply the updated chart:

```bash
helm upgrade -i langsmith langchain/langsmith \
  --values langsmith_config.yaml \
  --version <version> \
  --namespace <namespace> \
  --wait
```

### Enable with Terraform

The LangSmith Terraform modules can provision the required AWS and GCP infrastructure and generate the corresponding Helm values.

#### AWS

In `modules/aws/infra/terraform.tfvars`, enable Sandboxes and configure the sandbox node capacity:

```hcl
enable_sandboxes = true
redis_source     = "external"

sandbox_juicefs_redis_instance_type             = "cache.m6g.large"
sandbox_juicefs_redis_snapshot_retention_limit  = 7

sandbox_host_node_count               = 1
sandbox_host_instance_types           = ["m5d.metal"]
sandbox_host_configure_instance_store = true

sandbox_host_image_tag = "<same-release-tag-as-your-langsmith-images>"
```

> AWS Sandboxes require `redis_source = "external"`. The Terraform module creates a dedicated ElastiCache Redis instance for JuiceFS sandbox metadata, configures that dedicated instance with the recommended `noeviction` policy, reuses the LangSmith S3 bucket for sandbox object storage, creates the JuiceFS CSI config Secret, and adds the expected node label and taint.

Reviewer 2: ```suggestion
AWS Sandboxes require `redis_source = "external"`. The Terraform module:

- Creates a dedicated ElastiCache Redis instance for JuiceFS sandbox metadata. 
- Configures that dedicated instance with the recommended `noeviction` policy. 
- Reuses the LangSmith S3 bucket for sandbox object storage.
- Creates the JuiceFS CSI config Secret.
- Adds the expected node label and taint.
```

If you deploy the Helm release with the Terraform app module, set the sandbox app values in `modules/aws/app/terraform.tfvars` as well:

```hcl
enable_sandboxes      = true
chart_version          = "~0.16.0"
sandbox_host_image_tag = "<same-release-tag-as-your-langsmith-images>"
```

When `enable_sandboxes = true`, the Terraform app module requires an explicit LangSmith Helm chart version `0.16.0` or later and a sandbox runtime image tag.

Run the normal AWS flow:

```bash
make apply
make init-values
CHART_VERSION="~0.16.0" make deploy
```

#### GCP

In `modules/gcp/infra/terraform.tfvars`, enable Sandboxes and configure a Standard GKE node pool:

```hcl
enable_sandboxes = true
redis_source     = "external"

gke_use_autopilot     = false
enable_gcp_iam_module = true

sandbox_juicefs_redis_memory_size       = 5
sandbox_juicefs_redis_high_availability = true

sandbox_host_node_count     = 1
sandbox_host_min_node_count = 1
sandbox_host_max_node_count = 5
sandbox_host_machine_type   = "n2-standard-8"

sandbox_host_image_tag = "<same-release-tag-as-your-langsmith-images>"
```

> GCP Sandboxes require `redis_source = "external"`. The Terraform module creates a dedicated Memorystore Redis instance for JuiceFS sandbox metadata, configures that dedicated instance with the recommended `noeviction` policy, reuses the LangSmith GCS bucket for sandbox object storage, creates the JuiceFS CSI config Secret, and adds the expected node label and taint.

Reviewer 2: ```suggestion
GCP Sandboxes require `redis_source = "external"`. The Terraform module:

- Creates a dedicated Memorystore Redis instance for JuiceFS sandbox metadata. 
- Configures that dedicated instance with the recommended `noeviction` policy. 
- Reuses the LangSmith GCS bucket for sandbox object storage. 
- Creates the JuiceFS CSI config Secret. 
- Adds the expected node label and taint.
```
